Let's start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn_list = list(read_file)
hn = hn_list[1:]
headers = hn_list[0]
# Displaying first 5 rowe of the hn
print(hn_list[0:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Extracting first row of above result and storing it in headers variable as they are the headers of the dataset and rest all in variable hn

In [2]:
print(headers)
print(hn[0:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles

In [4]:
ask_posts = [ ]
show_posts = [ ]
other_posts = [ ]
for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    if title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(ask_posts[:5])
print(show_posts[:5])
print(len(ask_posts)) # number of ask HN posts  
print(len(show_posts)) # number of show HN posts
print(len(other_posts)) # length of other post

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h

## Calculating number of comments received on  'ask HN' posts
Let's calculate average nuber of comments received on ask HN posts

In [6]:
total_ask_comments = 0
for row in ask_posts:
    comments = int(row[4])
    total_ask_comments += comments
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)


14.038417431192661


## Calculating number of comments received on 'show HN' posts
Now we'll calculate average number of comments on show HN comments

In [7]:
total_show_comments = 0
for row in show_posts:
    comments = int(row[4])
    total_show_comments += comments
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


It is seen that average no. of comments asking for a solution for a problem that they are facing is more than the no.of comments that want to show what they have made or accomplished. It can seen that people are ready to help others by giving them solution .

In the last screen, you should've determined that, on average, ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
Calculate the average number of comments ask posts receive by hour created

In [9]:
import datetime as dt
result_list = [ ]
for row in ask_posts:
    created = row[6]
    comment = int(row[4])
    result_list.append([created,comment])
counts_by_hour = { }
comments_by_hour = { }
for row in result_list:
    date = row[0]
    com = row[1]
    hour = dt.datetime.strptime(date,"%m/%d/%Y %H:%M").strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = com
    if hour in counts_by_hour:
        counts_by_hour[hour] +=1
        comments_by_hour[hour] += com

Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.


In [10]:
avg_by_hour = [ ]
for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour] / counts_by_hour[hour]])
avg_by_hour

[['06', 8.844444444444445],
 ['08', 10.142857142857142],
 ['18', 13.1],
 ['07', 7.685714285714286],
 ['20', 21.28395061728395],
 ['15', 38.27350427350427],
 ['16', 16.798165137614678],
 ['10', 13.233333333333333],
 ['11', 10.898305084745763],
 ['17', 11.356435643564357],
 ['03', 7.672727272727273],
 ['23', 7.884057971014493],
 ['14', 13.13888888888889],
 ['09', 5.586956521739131],
 ['00', 8.160714285714286],
 ['05', 10.48936170212766],
 ['01', 11.737704918032787],
 ['13', 14.906976744186046],
 ['22', 6.680555555555555],
 ['19', 10.72972972972973],
 ['04', 7.083333333333333],
 ['12', 9.337837837837839],
 ['02', 23.45762711864407],
 ['21', 15.9]]

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [11]:
swap_avg_by_hour =[ ]
for row in avg_by_hour:
    
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[8.844444444444445, '06'], [10.142857142857142, '08'], [13.1, '18'], [7.685714285714286, '07'], [21.28395061728395, '20'], [38.27350427350427, '15'], [16.798165137614678, '16'], [13.233333333333333, '10'], [10.898305084745763, '11'], [11.356435643564357, '17'], [7.672727272727273, '03'], [7.884057971014493, '23'], [13.13888888888889, '14'], [5.586956521739131, '09'], [8.160714285714286, '00'], [10.48936170212766, '05'], [11.737704918032787, '01'], [14.906976744186046, '13'], [6.680555555555555, '22'], [10.72972972972973, '19'], [7.083333333333333, '04'], [9.337837837837839, '12'], [23.45762711864407, '02'], [15.9, '21']]


In [69]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments")
for avg in sorted_swap[:5]:
    hour = avg[1]
    num = avg[0]
    time = dt.datetime.strptime(hour,"%H").strftime("%H:%m")
    print("{t}:00: {a:.2f} average comments per post".format(t = time,a = num))

Top 5 Hours for Ask Posts Comments
15:01:00: 38.27 average comments per post
02:01:00: 23.46 average comments per post
20:01:00: 21.28 average comments per post
16:01:00: 16.80 average comments per post
21:01:00: 15.90 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.27 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.



In [12]:
points_of_ask_posts = [ ]
for row in ask_posts:
    points = int(row[3])
    points_of_ask_posts.append(points)
avg_points_ask = sum(points_of_ask_posts) / len(ask_posts)
print(avg_points_ask)

15.061926605504587


In [13]:
points_of_show_posts = [ ]
for row in show_posts:
    points = int(row[3])
    points_of_show_posts.append(points)
avg_points_show = sum(points_of_show_posts) / len(show_posts)
print(avg_points_show)

27.555077452667813


In [14]:
r_created = { }
r_points = { }
for row in show_posts:
    created = row[6]
    points = int(row[3])
    hour = dt.datetime.strptime(created,"%m/%d/%Y %H:%M").strftime("%H")
    if hour in r_created:
        r_created[hour] += 1
        r_points[hour] += points
    else:
        r_created[hour] = 1
        r_points[hour] = points

avg_points_hour = [ ]
for row in r_created:
    avg_points_hour.append([row,r_points[row] / r_created[row]])
print(avg_points_hour[:11])

[['06', 23.4375], ['08', 15.264705882352942], ['18', 36.31147540983606], ['07', 19.0], ['20', 30.316666666666666], ['11', 33.63636363636363], ['16', 28.322580645161292], ['10', 18.916666666666668], ['15', 28.564102564102566], ['17', 27.107526881720432], ['03', 25.14814814814815]]


In [15]:
swap_avg_points_hour = [ ]
for row in avg_points_hour:
    hour = row[0]
    points = row[1]
    swap_avg_points_hour.append([points,hour])
print(swap_avg_points_hour[:10])

[[23.4375, '06'], [15.264705882352942, '08'], [36.31147540983606, '18'], [19.0, '07'], [30.316666666666666, '20'], [33.63636363636363, '11'], [28.322580645161292, '16'], [18.916666666666668, '10'], [28.564102564102566, '15'], [27.107526881720432, '17']]


In [74]:
sorted_points = sorted(swap_avg_points_hour,reverse = True)
print(sorted_points[:10])

[[42.388888888888886, '23'], [41.68852459016394, '12'], [40.34782608695652, '22'], [37.83870967741935, '00'], [36.31147540983606, '18'], [33.63636363636363, '11'], [30.945454545454545, '19'], [30.316666666666666, '20'], [28.564102564102566, '15'], [28.322580645161292, '16']]


# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.